### ISSUE TITLE
Rewrite optimizations that effectively perform their own canonicalizations

### ISSUE THREAD

**Issue by brandonwillard** [aesara-devs/aesara#649]
Some optimizations perform their own implicit canonicalizations through complicated condition-checking logic; these optimizations need to be rewritten and new canonicalizations added.

For example, `aesara.tensor.math_opt.local_log1p` has logic for both `log(add(1, x))` and `log(sub(1, x))`.  Only one of those representations should be the canonical form.

As part of this issue, we need to do a survey to determine which optimizations have these kinds of assumptions, which canonicalizations are missing, which ones have logic for impossible cases, etc.

**Comment by kc611** [comment]
We should probably have a list of rewrites and maintain it, if we don't already. Something like that could even be put into the Documentation of the respective modules.

**Comment by brandonwillard** [comment]
> We should probably have a list of rewrites and maintain it, if we don't already. Something like that could even be put into the Documentation of the respective modules. It would make it easy for everyone to check at a glance which optimizations exists and which don't.
> 
> That said, our documentation for graph optimization can really use some updates: https://aesara.readthedocs.io/en/latest/extending/optimization.html

Yes, definitely.  It needs to be automated in some way, though.  We could parse all the entries in the canonicalizations DB and collect their docstrings, but this would require that we add and formalize the docstrings for every rewrites/optimizations, and it would still only be good for documentation purposes.

Ultimately, we need something for more than just documentation, but for programmatic introspection and analysis.  I've mentioned this before, but we really need to do some basic things like [rewrite completion](https://en.wikipedia.org/wiki/Knuth%E2%80%93Bendix_completion_algorithm).  In other words, we need to better understand our current set of rewrites (e.g. are they inconsistent in some cases), "optimize" them (e.g. add missing rewrites that prevent us from reaching canonical forms, simplify rewrites so that fewer steps are needed, etc.), and determine some basic properties of proposed additions _before_ adding them (e.g. do they lead to divergences).  Plus, we need to start working on our overall rewrite "strategies" (e.g. how, when, and in what order we perform sets of rewrites).

These things generally involve a higher level of systematization than we currently have.  The closest thing we have now is `PatternSub`, but it doesn't have the "expressiveness" necessary. 
 
Our planned use of miniKanren (e.g. via `kanren` in #523) is one way to get there, since the rewrites are then explicit S-expressions that can be easily parsed and manipulated in Python.  [Here's](https://github.com/brandonwillard/mk-rewrite-completion) an example involving rewrite completion from rewrites expressed in miniKanren.
